include/swift/AST/Evaluator.h

```diff
@@ -0,0 +1,264 @@
+//
+// This file defines the Evaluator class that evaluates and caches
+// requests.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_AST_EVALUATOR_H
+#define SWIFT_AST_EVALUATOR_H
+
+#include "swift/AST/AnyRequest.h"
+#include "swift/Basic/AnyValue.h"
+#include "swift/Basic/Defer.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/Optional.h"
+#include <tuple>
+#include <type_traits>
+#include <vector>
+
+namespace swift {
+
+using llvm::Optional;
+using llvm::None;
+
+class DiagnosticEngine;
+
+/// Evaluation engine that evaluates and caches "requests", checking for cyclic
+/// dependencies along the way.
+///
+/// Each request is a function object that accepts a reference to the evaluator
+/// itself (through which it can request other values) and produces a
+/// value. That value can then be cached by the evaluator for subsequent access,
+/// using a policy dictated by the request itself.
+///
+/// The evaluator keeps track of all in-flight requests so that it can detect
+/// and diagnose cyclic dependencies.
+///
+/// Each request should be its own function object, supporting the following
+/// API:
+///
+///   - Copy constructor
+///   - Equality operator (==)
+///   - Hashing support (hash_value)
+///   - TypeID support (see swift/Basic/TypeID.h)
+///   - The output type (described via a nested type OutputType), which
+///     must itself by a value type that supports TypeID.
+///   - Evaluation via the function call operator:
+///       OutputType operator()(Evaluator &evaluator) const;
+///   - Cycle breaking and diagnostics operations:
+///
+///       void diagnoseCycle(DiagnosticEngine &diags) const;
+///       void noteCycleStep(DiagnosticEngine &diags) const;
+///       OutputType breakCycle() Const;
+///   - Caching policy:
+///
+///     static const bool isEverCached;
+///
+///       When false, the request's result will never be cached. Note that this
+///       also precludes cycle detection.  When true, the result will be
+///       cached on completion. How it is cached depends on the following.
+///
+///     bool isCached() const;
+///
+///       Dynamically indicates whether to cache this particular instance of the
+///       request, so that (for example) requests for which a quick check
+///       usually suffices can avoid caching a trivial result.
+///
+///     static const bool hasExternalCache;
+///
+///       When false, the results will be cached within the evaluator and
+///       cannot be accessed except through the evaluator. This is the
+///       best approach, because it ensures that all accesses to the result
+///       are tracked.
+///
+///       When true, the request itself must provide an way to cache the
+///       results, e.g., in some external data structure. External caching
+///       should only be used when staging in the use of the evaluator into
+///       existing mutable data structures; new computations should not depend
```

Comment:
Hm! Are we going to migrate to, say, every `TypeLoc` going through the `Evaluator`? If so, it kinda seems like we may want something with a more continuous memory layout? I'm thinking of basically doing an AoS-to-SoA transform, something like:

```cpp
class Evaluator {
    // ...

    llvm::DenseMap<TypeId, AnyCache> caches;
    // ...
}

class AnyCache {
    // a type erased `llvm::DenseMap<Request, Request::OutputType>`
}
```

or, even, going fully-static/killing our compile times:

```cpp
template<typename Request...>
class Evaluator {
    std::tuple<llvm::DenseMap<Request, Request::OutputType>...> caches;
    // or whatever the template magic to get a DenseMap for each request type is.
    // (plus extra magic to elide the map for externally cached requests.)
}

using SwiftEvaluator = Evaluator< InheritedTypeRequest, SuperclassTypeRequest, ...>;
```

The `getResultCached` function would then first need to find the appropriate cache (a downside of the first approach is, of course, that this may be moderately expensive and so outweighs the benefits of that style vs. the `[AnyRequest : AnyValue?]` version), but I think the rest is the same.

Reply:
This is a fantastic idea! Not the fully-static one, which would kill our layering and our compile times, but the type-erased `AnyCache` would be much better.